datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -592,109 +591,51 @@ impl ExprMask {
     input_schema: DFSchemaRef,
-    // inner states
-    visit_stack: Vec<VisitRecord>,
-    /// increased in fn_down, start from 0.
-    node_count: usize,
     /// which expression should be skipped?
     expr_mask: ExprMask,
 }
 
-/// Record item that used when traversing a expression tree.
-enum VisitRecord {
-    /// `usize` is the monotone increasing series number assigned in pre_visit().
-    /// Starts from 0. Is used to index the identifier array `id_array` in post_visit().
-    EnterMark(usize),
-    /// the node's children were skipped => jump to f_up on same node
-    JumpMark(usize),
-    /// Accumulated identifier of sub expression.
-    ExprItem(Identifier),
-}
-
-impl ExprIdentifierVisitor<'_> {
-    /// Find the first `EnterMark` in the stack, and accumulates every `ExprItem`
-    /// before it.
-    fn pop_enter_mark(&mut self) -> (usize, Identifier) {
-        let mut desc = String::new();
-
-        while let Some(item) = self.visit_stack.pop() {
-            match item {
-                VisitRecord::EnterMark(idx) | VisitRecord::JumpMark(idx) => {
-                    return (idx, desc);
-                }
-                VisitRecord::ExprItem(id) => {
-                    desc.push_str(&id);
-                }
-            }
-        }
-        unreachable!("Enter mark should paired with node number");
-    }
-}
-
 impl TreeNodeVisitor for ExprIdentifierVisitor<'_> {
     type Node = Expr;
 
     fn f_down(&mut self, expr: &Expr) -> Result<TreeNodeRecursion> {
         // related to https://github.com/apache/datafusion/issues/8814
         // If the expr contain volatile expression or is a short-circuit expression, skip it.
         if expr.short_circuits() || expr.is_volatile()? {
-            self.visit_stack
-                .push(VisitRecord::JumpMark(self.node_count));
             return Ok(TreeNodeRecursion::Jump); // go to f_up
         }
 
-        self.visit_stack
-            .push(VisitRecord::EnterMark(self.node_count));
-        self.node_count += 1;
-
         Ok(TreeNodeRecursion::Continue)
     }
 
     fn f_up(&mut self, expr: &Expr) -> Result<TreeNodeRecursion> {
-        let (_idx, sub_expr_identifier) = self.pop_enter_mark();
-
         // skip exprs should not be recognize.
         if self.expr_mask.ignores(expr) {
-            let curr_expr_identifier = ExprSet::expr_identifier(expr);
-            self.visit_stack
-                .push(VisitRecord::ExprItem(curr_expr_identifier));
             return Ok(TreeNodeRecursion::Continue);
         }
-        let curr_expr_identifier = ExprSet::expr_identifier(expr);
-        let alias_symbol = format!("{curr_expr_identifier}{sub_expr_identifier}");
 
-        self.visit_stack
-            .push(VisitRecord::ExprItem(alias_symbol.clone()));
+        let curr_expr_identifier = ExprSet::expr_identifier(expr);
 
         let data_type = expr.get_type(&self.input_schema)?;
 
+        let alias_symbol = format!("#{{{curr_expr_identifier}}}");
```

Comment:
A small nit might be to pull this into a function that can be documented (so the code that creates identifiers is easier to find)

Perhaps something like 
```rust
let alias_symbol = ExprSet::alias_symbol(expr)
```

🤔

Reply:
Love that, this can also help us create the alias on-the-fly instead of passing it inside the expr_set.